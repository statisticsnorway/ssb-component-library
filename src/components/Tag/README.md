# Tag

> Action Tag

### Usage

#### HTML

```html
<button class="ssb-tag">Tag</button>

<button class="ssb-tag">
  <div class="st-icon">
    <i>{feather.star 18px}</i>
  </div>
  Tag with icon
</button>
```

#### React

```jsx harmony
<Tag>Tag</Tag>

<Tag icon={<Star size={18} />}>Tag with icon</Tag>
```

Available props:

| Name      | Type          | Description              |
| --------- | ------------- | ------------------------ |
| children  | Required node | Tag text or/and icon     |
| className | string        | Optional container class |
| icon      | node          | Renders an icon          |
| onClick   | function      | Callback for tag click   |
