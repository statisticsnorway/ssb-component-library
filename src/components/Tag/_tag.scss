.ssb-tag {
  @include roboto;
  @include reset-button;
  align-items: center;
  background: $ssb-white;
  border: 1px solid $ssb-green-4;
  border-radius: 20px;
  color: $ssb-green-4;
  display: flex;
  font-size: 16px;
  font-weight: bold;
  height: 36px;
  line-height: 1.25;
  min-height: 26px;
  padding: 8px 16px;
  text-align: center;
  transition:
    background 0.18s,
    color 0.18s;
  white-space: nowrap;

  &:hover,
  &:active {
    cursor: pointer;
  }

  &:hover,
  &:focus {
    @include focus-ring;
    background: $ssb-green-4;
    color: $ssb-white;
  }

  &:active {
    background: $ssb-dark-5;
    border: 2px solid $ssb-dark-5;
    color: $ssb-white;
  }

  .st-icon {
    display: inline-flex;
    font-size: 18px;
    margin-right: 5px;
  }
}
