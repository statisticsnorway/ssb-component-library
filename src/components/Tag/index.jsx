import React from 'react';
import PropTypes from 'prop-types';

const Tag = ({
	children, className, icon, onClick,
}) => (
	<button
		className={`ssb-tag${className ? ` ${className}` : ''}`}
		onClick={onClick}
	>
		{icon && <div className="st-icon">{icon}</div>}
		{children}
	</button>
);

Tag.defaultProps = {
	className: '',
	onClick: () => {},
};

Tag.propTypes = {
	children: PropTypes.node.isRequired,
	className: PropTypes.string,
	icon: PropTypes.node,
	onClick: PropTypes.func,
};

export default Tag;
