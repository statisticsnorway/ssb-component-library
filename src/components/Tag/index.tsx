import React, { type ReactNode } from 'react'

interface TagProps {
  children: ReactNode
  className?: string
  icon?: ReactNode
  onClick?: () => void
}

const Tag: React.FC<TagProps> = ({ children, className, icon, onClick = () => {} }) => (
  <button className={`ssb-tag${className ? ` ${className}` : ''}`} onClick={onClick}>
    {icon && <div className='st-icon'>{icon}</div>}
    {children}
  </button>
)

export default Tag
