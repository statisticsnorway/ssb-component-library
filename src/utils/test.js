import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
