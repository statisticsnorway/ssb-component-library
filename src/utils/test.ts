import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
