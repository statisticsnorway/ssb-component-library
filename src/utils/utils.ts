// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
