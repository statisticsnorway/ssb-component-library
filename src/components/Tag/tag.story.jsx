import React from 'react'
import { Star } from 'react-feather'
import Tag from './index'

export default {
  title: 'Tag',
}

export const Default = () => (
  <div>
    <Tag>Tag</Tag>
  </div>
)

export const TagWithIcon = () => (
  <div>
    <Tag icon={<Star size={18} />}>Tag with icon</Tag>
  </div>
)

TagWithIcon.story = {
  name: 'Tag with icon',
}
