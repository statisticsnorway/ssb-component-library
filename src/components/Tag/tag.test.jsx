import React from 'react'
import userEvent from '@testing-library/user-event'

import { screen, render } from '../../utils/test'
import Tag from './index'

describe('Tag component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(<Tag>Tag</Tag>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Toggles classNames correctly', () => {
    const { asFragment } = render(<Tag className='mr-4'>Tag</Tag>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Renders icon', () => {
    const { asFragment } = render(<Tag icon={<i />}>Tag</Tag>)
    expect(asFragment()).toMatchSnapshot()
  })
  test('Sends callback', async () => {
    const onClick = jest.fn()
    const user = userEvent.setup()

    render(<Tag onClick={onClick}>Tests</Tag>)
    const button = screen.getByText('Tests')
    await user.click(button)
    expect(onClick).toBeCalled()
    await user.click(button)
    expect(onClick).toBeCalledTimes(2)
  })
})
