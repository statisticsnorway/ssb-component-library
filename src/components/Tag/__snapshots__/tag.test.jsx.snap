// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tag component Matches the snapshot 1`] = `
<DocumentFragment>
  <button
    class="ssb-tag"
  >
    Tag
  </button>
</DocumentFragment>
`;

exports[`Tag component Renders icon 1`] = `
<DocumentFragment>
  <button
    class="ssb-tag"
  >
    <div
      class="st-icon"
    >
      <i />
    </div>
    Tag
  </button>
</DocumentFragment>
`;

exports[`Tag component Toggles classNames correctly 1`] = `
<DocumentFragment>
  <button
    class="ssb-tag mr-4"
  >
    Tag
  </button>
</DocumentFragment>
`;
